File: .NET/Microsoft.Recognizers.Text.Choice/Utilities/UnicodeUtils.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Globalization;

namespace Microsoft.Recognizers.Text.Choice.Utilities
{
    using System.Collections.Generic;

    public static class UnicodeUtils
    {
        public static bool IsEmoji(string letter)
        {
            const int whereEmojiLive = 0xFFFF; // Supplementary Unicode Plane. This is where emoji live
            return char.IsHighSurrogate(letter[0]) && char.ConvertToUtf32(letter, 0) > whereEmojiLive;
        }

        public static IEnumerable<string> Letters(string text)
        {
            char? codePoint = null;
            foreach (char c in text)
            {
                if (codePoint != null)
                {
                    yield return new string(new[] { codePoint.Value, c });
                    codePoint = null;
                }
                else if (!char.IsHighSurrogate(c))
                {
                    yield return c.ToString(CultureInfo.InvariantCulture);
                }
                else
                {
                    codePoint = c;
                }
            }
        }
    }
}

File: Java/libraries/recognizers-text-choice/src/main/java/com/microsoft/recognizers/text/choice/utilities/UnicodeUtils.java
package com.microsoft.recognizers.text.choice.utilities;

import java.lang.Character;
import java.util.ArrayList;
import java.util.List;

public class UnicodeUtils {
    public static boolean isEmoji(String letter) {
        final int WhereEmojiLive = 0xFFFF; // Supplementary Unicode Plane. This is where emoji live
        return Character.isHighSurrogate(letter.charAt(0)) && (letter.codePoints().sum() > WhereEmojiLive);
    }

    public static List<String> letters(String text) {
        char codePoint = 0;
        List<String> result = new ArrayList<>();
        for (int i = 0; i < text.length(); i++) {
            char c = text.charAt(i);
            if (codePoint != 0) {
                result.add(new String(Character.toChars(codePoint + c)));
                codePoint = 0;
            } else if (!Character.isHighSurrogate(c)) {
                result.add(Character.toString(c));
            } else {
                codePoint = c;
            }
        }
        return result;
    }
}